### ISSUE TITLE
Ability to whitelist certain users

### ISSUE THREAD

**Issue by fl0cke** [scheb/2fa#49]
Hey,

I'd like to have the ability to exclude certain users or user groups from two factor authentication.
There is already a PreferredProviderInterface, which could be extended to skip two factor authentication for users where `getPreferredTwoFactorProvider()` returns `null`. Another possible approach would be to provide a more generic `TwoFactorAuthWhitelistProvider` interface, which gets passed the authentication context and then decides if two factor auth is required or not:

```
interface TwoFactorAuthWhitelistProvider
{
    /*
     * Return false to exempt this login request from two factor authentication.
     */
    function shouldPerformTwoFactorAuth(AuthenticationContextInterface $context): bool;
}
```

This is basically a generalization of the `IpWhitelistProviderInterface`.

What do you think?

**Comment by scheb** [comment]
I'm not really sure if I get that right. If you want certain users to be excluded from 2fa, why do you let them activate 2fa?

Also, every 2fa provider has a method to control whether 2fa should be executed or not for a user. For example, if you use the email authentication provider, you have to implement the interface `Scheb\TwoFactorBundle\Model\Email\TwoFactorInterface` in your user entity and that interface has a method `isEmailAuthEnabled`:

https://github.com/scheb/2fa/blob/a2201c5df2618d3d5ca3f14e78e3308c7478b070/src/email/Model/Email/TwoFactorInterface.php#L9-L12

So, if you want to exclude a user from 2fa, you make that method return `false`.

Or am I missing something about your use-case? 🤔

**Comment by fl0cke** [comment]
Yes, that's right, But i don't want to disable 2fa for that user _completely_, just in certain conditions (similar to the "trusted device" or "ip whitelisting" feature). What i'm proposing here is just a generalization of those feature that is a more flexible and customizable by the user (me).

**Comment by scheb** [comment]
Well, if all on-board features aren't enough for you and you cannot make the decision purely based on user data within the `is*AuthEnabled()` method, there's a neat trick you can do:

Implement your own "ip whitelist provider" based on `IpWhitelistProviderInterface`, but instead of returning a list of IPs, you put your decision logic there. Whenever the user should be excluded from 2fa, you return the IP of the *current request* - if not, you return an empty list. That way, the current request (and user) can be excluded from 2fa.

Does that work for you?

My feeling is, there's no significant need for even more flexibility, at least not enough to justify putting in another exclusion layer. I might be wrong about the feeling, but until today no one brought that up to me. **If anyone else reads this and has the same need, please +1 this issue to make your need visible :)**

**Comment by fl0cke** [comment]
Yeah, i actually did something similar already: I implemented my own handler based on the `AuthenticationHandlerInterface` to decorate/wrap the existing ones. But that felt quite hacky, so i came up with this proposal to smooth things out a little ;) However, i understand your concerns, so maybe this could be something for the next major release?!

**Comment by scheb** [comment]
> Yeah, i actually did something similar already: I implemented my own handler based on the `AuthenticationHandlerInterface` to decorate/wrap the existing ones. But that felt quite hacky, so i came up with this proposal to smooth things out a little ;)

Yea, adding your own handler works as well, but if you're hooking into internal implementation like that, you have to accept the risk that things might break. I can't give you any guarantees that this handler solution will work in future releases (even within the 5.x branch). Using the IP whitelist for that purpose would be the more stable solution, because it's made to be an extension/customization point and therefore will not change.

> However, i understand your concerns, so maybe this could be something for the next major release?!

Maybe, but only if there's really a need for more flexibility that would justify any breaking changes.

**Comment by scheb** [comment]
Found a way to implement it in a non-breaking way, you can now inject custom conditions for two-factor authentication. See: https://github.com/scheb/2fa/blob/5.x/doc/custom_conditions.md

Released in version 5.7.0.
